lib/temporalio/worker/workflow_worker.rb

```diff
@@ -0,0 +1,127 @@
+require 'temporal/sdk/core/workflow_commands/workflow_commands_pb'
+require 'temporalio/error/failure'
+require 'temporalio/errors'
+require 'temporalio/interceptor'
+require 'temporalio/interceptor/chain'
+require 'temporalio/worker/workflow_runner'
+
+module Temporalio
+  class Worker
+    # @api private
+    class WorkflowWorker
+      def initialize(namespace, task_queue, worker, workflows, converter, interceptors)
+        @namespace = namespace
+        @task_queue = task_queue
+        @worker = worker
+        @workflows = prepare_workflows(workflows)
+        @converter = converter
+        @inbound_interceptors = Temporalio::Interceptor::Chain.new(
+          Temporalio::Interceptor.filter(interceptors, :workflow_inbound),
+        )
+        @outbound_interceptors = Temporalio::Interceptor::Chain.new(
+          Temporalio::Interceptor.filter(interceptors, :workflow_outbound),
+        )
+        @drain_queue = Queue.new
+        @running_workflows = {}
+      end
+
+      def run(reactor)
+        # @type var outstanding_tasks: Array[Async::Task]
+        outstanding_tasks = []
+
+        loop do
+          workflow_activation = worker.poll_workflow_activation
+          outstanding_tasks << reactor.async do |async_task|
+            handle_activation(workflow_activation)
+          ensure
+            outstanding_tasks.delete(async_task)
+          end
+        end
+      rescue Temporalio::Bridge::Error::WorkerShutdown
+        # No need to re-raise this error, it's a part of a normal shutdown
+      ensure
+        outstanding_tasks.each(&:wait)
+        drain_queue.close
+      end
+
+      def drain
+        drain_queue.pop
+      end
+
+      private
+
+      attr_reader :namespace, :task_queue, :worker, :workflows, :converter, :inbound_interceptors,
+                  :outbound_interceptors, :drain_queue, :running_workflows
+
+      def prepare_workflows(workflows)
+        workflows.each_with_object({}) do |workflow, result|
+          unless workflow.ancestors.include?(Temporalio::Workflow)
```

Comment:
Why not `workflow.class.is_a?(Temporalio::Workflow)`?

Reply:
In Ruby any class is an instance of a `Class` class:

```ruby
>> String.is_a?(String)
=> false
>> String.is_a?(Class)
=> true
```

Therefore we need to check the ancestor chain on a class to see if it's of a required type